sdk/inc/azure/iot/az_iot_pnp_client.h

```diff
@@ -0,0 +1,730 @@
+ */
+#define AZ_IOT_PNP_CLIENT_WEB_SOCKET_PATH "/$iothub/websocket"
+
+/**
+ * @brief The HTTP URI Path necessary when connecting to IoT Hub using WebSockets without an X509
+ * client certificate.
+ * @note Most devices should use #AZ_IOT_PNP_CLIENT_WEB_SOCKET_PATH. This option is available for
+ * devices not using X509 client certificates that fail to connect to IoT Hub.
+ */
+#define AZ_IOT_PNP_CLIENT_WEB_SOCKET_PATH_NO_X509_CLIENT_CERT \
+  AZ_IOT_PNP_CLIENT_WEB_SOCKET_PATH "?iothub-no-client-cert=true"
+
+/**
+ * @brief Gets the MQTT user name.
+ *
+ * The user name will be of the following format:
+ *
+ * `{iothubhostname}/{device_id}/?api-version=2020-09-30&{user_agent}&model-id={model_id}`
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[out] mqtt_user_name A buffer with sufficient capacity to hold the MQTT user name.
+ *                            If successful, contains a null-terminated string with the user name
+ *                            that needs to be passed to the MQTT client.
+ * @param[in] mqtt_user_name_size The size, in bytes of \p mqtt_user_name.
+ * @param[out] out_mqtt_user_name_length __[nullable]__ Contains the string length, in bytes, of
+ *                                                      \p mqtt_user_name. Can be `NULL`.
+ *
+ * @pre \p client must not be `NULL`.
+ * @pre \p mqtt_user_name must not be `NULL`.
+ * @pre \p mqtt_user_name_size must be greater than 0.
+ *
+ * @return An #az_result value indicating the result of the operation.
+ */
+AZ_NODISCARD AZ_INLINE az_result az_iot_pnp_client_get_user_name(
+    az_iot_pnp_client const* client,
+    char* mqtt_user_name,
+    size_t mqtt_user_name_size,
+    size_t* out_mqtt_user_name_length)
+{
+  return az_iot_hub_client_get_user_name(
+      &(client->_internal.iot_hub_client),
+      mqtt_user_name,
+      mqtt_user_name_size,
+      out_mqtt_user_name_length);
+}
+
+/**
+ * @brief Gets the MQTT client id.
+ *
+ * The client id will be of the following format:
+ *
+ * `{device_id}`
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[out] mqtt_client_id A buffer with sufficient capacity to hold the MQTT client id.
+ *                            If successful, contains a null-terminated string with the client id
+ *                            that needs to be passed to the MQTT client.
+ * @param[in] mqtt_client_id_size The size, in bytes of \p mqtt_client_id.
+ * @param[out] out_mqtt_client_id_length __[nullable]__ Contains the string length, in bytes, of
+ *                                                      of \p mqtt_client_id. Can be `NULL`.
+ *
+ * @pre \p client must not be `NULL`.
+ * @pre \p mqtt_client_id must not be `NULL`.
+ * @pre \p mqtt_client_id_size must be greater than 0.
+ *
+ * @return An #az_result value indicating the result of the operation.
+ */
+AZ_NODISCARD AZ_INLINE az_result az_iot_pnp_client_get_client_id(
```

Comment:
> If we were to remove the inline because the implementation has to change, would that be a breaking change?

I don't think removing inline is breaking, but this approach is impacting the public header files, only because of something that is an implementation detail. Therefore, it isn't a critical issue. **Your call.**

We have some specific places where we do inline methods, az_span_create being a canonical one. What makes that different is:
1) The implementation/number of fields/etc. is very unlikely to change (well-trodden design following precedence from a bunch of other languages)
2) It is used a lot of times in the SDK and in tight loops, and the implementation is very small/simple, so giving the compiler this hint is fruitful.

Though we don't optimize our SDK for this, as a concrete downside of inlining too liberally, see: https://github.com/Azure/azure-sdk-for-c/issues/359

Reply:
updated to remove the inline 👍